### PR TITLE
#30 Improved compatibility with other plugins that patch the POM for `provided` scope

### PR DESCRIPTION
**Pull request by dnault**
Treat `provided` scope as a synonym for `runtime` when performing dynamic version resolution. If an unrecognized scope is encountered, skip version resolution for that dependency instead of failing to generate the POM.

With this change to ResolvedMavenPlugin, `nebula.provided-base` and `nebula.nebula-maven-publishing` may be applied in any order with respect to each other, and the publishing plugin will resolve dynamic versions in `provided` dependencies.

### CODE REVIEW

File: src/test/groovy/nebula/plugin/publishing/maven/ResolvedMavenPluginIntSpec.groovy [hunk 1]
```diff
@@ -58,6 +58,37 @@ class ResolvedMavenPluginIntSpec extends IntegrationSpec {
         asmDep.version.text() == '2.2.3'
     }
 
+    def 'compatible with nebula.provided-base plugin'() {
+        writeHelloWorld('nebula.hello')
+        buildFile << '''
+            buildscript {
+                repositories { jcenter() }
+                dependencies { classpath "com.netflix.nebula:gradle-extra-configurations-plugin:2.2.0" }
```
Comment (bmuschko) [inline src/test/groovy/nebula/plugin/publishing/maven/ResolvedMavenPluginIntSpec.groovy]: I think we will want to be able to test this without the direct dependency to the plugin. Instead we can just create a new configuration named `provided`.

File: src/test/groovy/nebula/plugin/publishing/maven/ResolvedMavenPluginIntSpec.groovy [hunk 2]
```diff
@@ -58,6 +58,37 @@ class ResolvedMavenPluginIntSpec extends IntegrationSpec {
...
+            }
+            apply plugin: 'java'
+            apply plugin: "nebula.provided-base"
+            apply plugin: 'nebula-maven-publishing'
+            repositories { jcenter() }
```
Comment (bmuschko) [inline src/test/groovy/nebula/plugin/publishing/maven/ResolvedMavenPluginIntSpec.groovy]: The nebula-test library provides a way to publish and reference libraries locally without having to reach out to JCenter. We will want to do this here. I believe the extra-configurations plugin code has plenty of examples on how to use it.

**Comment by bmuschko** [comment]
Thanks for the pull request. I added some comments. Would you mind touching up the code? Let me know if you have questions.

File: src/test/groovy/nebula/plugin/publishing/maven/ResolvedMavenPluginIntSpec.groovy (hunk 1 above)
Comment (dnault) [inline src/test/groovy/nebula/plugin/publishing/maven/ResolvedMavenPluginIntSpec.groovy]: I'm glad you suggested this; interestingly, just adding a `provided` configuration (and having `compile` extend from it) isn't sufficient to reproduce the failure. The POM patching performed by the `nebula.provided-base` plugin appears to be what disrupts `ResolvedMavenPlugin`. A workaround for the issue is to apply the `nebula.provided-base` plugin *after* `nebula.nebula-maven-publishing`.

**Comment by dnault** [comment]
@bmuschko Thanks for the feedback. Explicit references to the `nebula.provided-base` plugin have been removed. Instead, the integration test emulates the behavior of that plugin. I investigated extending PluginProjectSpec instead of IntegrationSpec, but frankly I was a bit overwhelmed. For what it's worth, the test case added in this updated pull request is very similar to existing tests in the same file. What do you think?

**Comment by bmuschko** [comment]
Thanks for putting in the work. I added some more comments. We can merge this after you fix them.

**Comment by dnault** [comment]
Removed the semicolon. Modified the test so it:

- No longer prints the generated POM to stdout
- Uses DependencyGenerator instead of calling out to jCenter  
- Defines a custom configuration
- Is named "Can resolve dynamic versions after another plugin has patched the POM to support provided scope"

Any additional feedback / suggestions for better test name?

File: src/main/groovy/nebula/plugin/publishing/maven/ResolvedMavenPlugin.groovy [hunk 3]
```diff
@@ -51,6 +51,10 @@ class ResolvedMavenPlugin implements Plugin<Project> {
                         [scope, resolutionMap]
                     }
 
+                    // Another plugin may have already patched the POM to support 'provided' scope.
+                    // For the purpose of version resolution, 'provided' scope is equavalent to 'runtime'.
+                    scopeResolutionMap['provided'] = scopeResolutionMap['runtime']
```
Comment (rspieldenner) [inline src/main/groovy/nebula/plugin/publishing/maven/ResolvedMavenPlugin.groovy]: Is this going to be ordering dependent? If another plugin configures provided in this Map will the ordering the plugins are applied matter?

Comment (dnault) [inline src/main/groovy/nebula/plugin/publishing/maven/ResolvedMavenPlugin.groovy]: The order of plugin application will not matter, unless the other plugin does something exotic like adding a dependency to the `runtime` configuration after the project has been evaluated (in which case the `maven-publish` plugin doesn't even include the dependency in the POM, so it shouldn't really be an issue).

With this change, `ResolvedMavenPlugin` assumes any dependencies marked as `<scope>provided</scode>` in the POM are members of the `runtime` configuration, which is likely to be true in common usage. If a certain plugin implements `provided` differently, the worst that happens is that the nebula plugin will not resolve dynamic versions for the provided dependencies.

**Pull request merged**